vowpalwabbit/core/src/learner.cc

```diff
@@ -312,77 +326,427 @@ bool ec_is_example_header(const example& ec, label_type_t label_type)
   return false;
 }
 
-}  // namespace LEARNER
-}  // namespace VW
+learner* as_multiline(learner* l)
+{
+  if (l->is_multiline()) { return l; }
+  auto message = fmt::format("Tried to use a singleline reduction as a multiline reduction Name: {}", l->get_name());
+  THROW(message);
+}
 
-void VW::LEARNER::details::learner_build_diagnostic(VW::string_view this_name, VW::string_view base_name,
-    prediction_type_t in_pred_type, prediction_type_t base_out_pred_type, label_type_t out_label_type,
-    label_type_t base_in_label_type, details::merge_fn merge_fn_ptr, details::merge_with_all_fn merge_with_all_fn_ptr)
+learner* as_singleline(learner* l)
```

Comment:
Because these no longer actually change the type I don't think it should be called `as_X` anymore. If we are checking an invariant then we should rename to something like `require_singleline`